fix(gnovm): allow multiple NaN keys in map

thehowl opened the pull request:
This PR fixes #3867. It's an alternative to #3932. Depends on #4327.

- [As Jae requested,](https://github.com/gnolang/gno/pull/4114#issuecomment-3043155234) ComputeMapKey now returns a second value indicating whether a NaN is contained in the map key, and as such the map key is != to all others.
- Negative 0 is now also normalized to `0`, matching go's behaviour.
- Map keys are now re-assigned when using `GetPointerForKey` on an existing key, in order to match [Go's behaviour](https://go.dev/play/p/iNPDR4FQlRv).
- It detects NaN using the `softfloat` package, rather than using hardware FP instructions, ensuring cross-architecture determinism.

This PR doesn't strictly depend on #4327, also because instead of copying the key in `GetPointerForKey` we copy it in its users, allowing to call `DidUpdate` on the copied key as necessary, however it does work from the same "fix".

Gno2D2 commented:
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: thehowl/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

thehowl commented:
There is a possible problem: this behaves differently in Go and Gno...

```go
// You can edit this code!
// Click here and start typing.
package main

import (
	"fmt"
	"math"
)

func main() {
	n0 := math.Float64frombits(1 << 63)
	fmt.Println(n0)
	m := map[float64]string{}
	m[0] = "hey"
	m[n0] = "world"
	fmt.Println(m)
}

// Output:
// -0
// map[0:world]
```

In Go, the key that is used is actually the last: https://go.dev/play/p/K--afH1wh0L -- so the result is `map[-0:world]`.

thehowl commented:
ALSO realised this doesn't work for ie. a struct with a float... we probably still need some modification to ComputeMapKey

mvertes reviewed (approved):
Looks good to me

thehowl commented:
I replied to this review on Signal a while back, cross-posting here:

![image](https://github.com/user-attachments/assets/f6f50783-28b3-4904-ab1a-60bff9ae9172)


<details><summary>Text-based version, for searchability</summary>
<p>


> The idea behind the PR was to follow in the footsteps of what Marc did, which I think largely makes sense
> 
> And that is not to change the code path, ie. adding conditionals or slowing down code anyway, for the 99.9999% of cases that don't have NaNs
> 
> Also keep in mind any solution has to work not just for map[float64]V, but also for map[struct{k float64}]V
> 
> So you need a "deep" nan check for those cases, which is potentially expensive for large structs.
> You can make this cheaper by returning a second param in ComputeMapKey, but again, this makes code more complex to account for the 1/1000000th case where there is a nan
> 
> because then you need to add a check for the second param each time

</p>
</details>

jaekwon commented on gnovm/pkg/gnolang/values.go:
```diff
@@ -725,6 +727,11 @@ func (mv *MapValue) GetLength() int {
+		// When assigning to a map item, the key is always equal to that of the
+		// last assignment; this is mostly noticeable in the case of -0 / 0:
+		// https://go.dev/play/p/iNPDR4FQlRv
+		mli.Key = key2
```
i don't see why mli needs to be modified here.

ltzmaxwell replied:
cross reference: https://github.com/gnolang/gno/pull/4327

thehowl commented on gnovm/pkg/gnolang/values.go:
```diff
@@ -1052,64 +1072,76 @@ func (tv TypedValue) unrefCopy(alloc *Allocator, store Store) (cp TypedValue) {
 // Returns encoded bytes for primitive values.
-// These bytes are used for both value hashes as well
-// as hash key bytes.
-func (tv *TypedValue) PrimitiveBytes() (data []byte) {
```
There are no other uses at present. We can always add it back when necessary.

ltzmaxwell commented on gnovm/pkg/gnolang/values.go:
```diff
@@ -724,12 +725,19 @@ func (mv *MapValue) GetLength() int {
+	kmk, isNaN := key.ComputeMapKey(store, false)
+	if !isNaN {
+		if mli, ok := mv.vmap[kmk]; ok {
+			key2 := key.Copy(alloc)
```
can you please check this PR: https://github.com/gnolang/gno/pull/4327. 
it looks `key` should always be copied?

thehowl replied:
Okay, sounds good. I'll apply that change here with you as co-author.

thehowl commented on gnovm/pkg/gnolang/values.go:
```diff
@@ -1950,10 +1990,13 @@ func (tv *TypedValue) GetPointerAtIndex(rlm *Realm, alloc *Allocator, store Stor
-		key := iv.ComputeMapKey(store, false)
-		if _, ok := mv.vmap[key]; ok {
-			exist = true
+		var oldObject Object
+		key, isNaN := iv.ComputeMapKey(store, false)
+		if !isNaN {
+			k, ok := mv.vmap[key]
+			if ok {
+				oldObject = k.Key.GetFirstObject(store)
+			}
```
exactly; it follows from the other change in the realm logic.

Updating the comment

thehowl commented:
I addressed maxwell's review, and updated the OP to match the changed method.

I suggest we merge #4327 first, and then go ahead and merge this - @ltzmaxwell, in the meantime, if you could have a second look and tell me if the changes are sensible ;)
